File: Interview/src/main/java/com/singerstone/jojo/打家劫舍123.java
package com.singerstone.jojo;

import static com.singerstone.jojo.二叉树构建.makeTreeRecursion;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

public class 打家劫舍123 {

    public static void main(String[] args) {
        Integer[] values = {0, 1, 2, 3, 4};
        // 0
        //1 2
        //34
        TreeNode root = makeTreeRecursion(values, 0, values.length);
        System.out.println(new 打家劫舍123().rob(root));

        int[] array = new int[]{1, 2, 3, 4, 5};
        System.out.println(new 打家劫舍123().rob1(array));
        System.out.println(new 打家劫舍123().rob2(array));

    }

    int rob1(int[] array) {
        int[] dp = new int[array.length];
        dp[0] = array[0];
        dp[1] = Math.max(array[0], array[1]);
        for (int i = 2; i < array.length; i++) {
            dp[i] = Math.max(dp[i - 1], array[i] + dp[i - 2]);
        }
        return dp[array.length - 1];
    }

    int rob2(int[] array) {
        int[] arr1 = Arrays.copyOfRange(array, 0, array.length - 1);
        int[] arr2 = Arrays.copyOfRange(array, 1, array.length);
        return Math.max(rob1(arr1), rob1(arr2));
    }


    Map<TreeNode, Integer> inMap = new HashMap<>();
    Map<TreeNode, Integer> outMap = new HashMap<>();

    int max = 0;

    //4,1,null,2,null,3
    public int rob(TreeNode root) {
        if (root == null) {
            inMap.put(null, 0);
            outMap.put(null, 0);
            return 0;
        }
        rob(root.left);
        rob(root.right);
        int inValue = root.value + outMap.get(root.left) + outMap.get(root.right);
        // 这里要注意的是：不rob当前的节点，并不代表一定要rob子节点，所以这里要取最大值
        // int outValue = Math.max(inMap.get(root.left) + inMap.get(root.right), outMap.get(root.left) + outMap.get(root.right));
        int outValue = Math.max(inMap.get(root.left), outMap.get(root.left)) + Math.max(
                inMap.get(root.right), outMap.get(root.right));
        inMap.put(root, inValue);
        outMap.put(root, outValue);
        max = Math.max(max, Math.max(inValue, outValue));
        // 返回当前
        return max;
    }

}

File: Interview/src/main/java/com/singerstone/jojo/打家劫舍123.kt
package com.singerstone.jojo

import kotlin.math.max

class 打家劫舍123 {

    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            val input = arrayOf(1, 2, 3, 4, 5)
            println(cashDp1(input))
            println(cashDp2(input))
            val array = arrayOf(0, 1, 2, 3, 4, 5, 6, 7, 8, 9)
            println(cashDp3(二叉树构建.makeTreeRecursion(array, 0, array.size), hashMapOf()))

        }

        // 打家劫舍 1
        // dp[i]=max(dp(i-1),arr[i]+dp[i-2])
        fun cashDp1(array: Array<Int>): Int {
            val dp = arrayOfNulls<Int>(array.size)
            dp[0] = array[0]
            dp[1] = max(array[0], array[1])
            for (i in 2 until array.size) {
                dp[i] = max(dp[i - 1]!!, array[i] + dp[i - 2]!!)
            }

            return dp[array.size - 1]!!


        }

        /**
         * 打家劫舍2 ，村民围成一个环.
         * 思路是分别计算去掉第一个和最后一个的最大收益
         * 及时分别去掉 1 ，5 计算最大值 但是这个是为啥呢？
         */
        fun cashDp2(array: Array<Int>): Int {
            var arr1 = array.copyOfRange(0, array.size - 1)
            var arr2 = array.copyOfRange(1, array.size)
            return max(cashDp1(arr1), cashDp1(arr2))
        }

        /**
         * 现在村民是树形结构
         *       1
         *     2   3
         *   4  5 6 7
         * 8  9
         *
         *
         */
        fun cashDp3(head: TreeNode?, record: HashMap<TreeNode, Int>): Int {
            if (head == null) {
                return 0
            }
            if (record.containsKey(head)) {
                return record[head]!!
            }
            var ll = 0
            var lr = 0
            var rl = 0
            var rr = 0
            head.left?.left?.let {
                ll = cashDp3(it, record)
            }
            head.left?.right?.let {
                lr = cashDp3(it, record)
            }
            head.right?.left?.let {
                rl = cashDp3(it, record)
            }
            head.right?.right?.let {
                rr = cashDp3(it, record)
            }
            val cur = max(head.value + ll + lr + rl + rr, cashDp3(head.left, record) + cashDp3(head.right, record))
            record[head] = cur
            return cur
        }

    }
}